src/dbnode/storage/series/buffer.go

```diff
@@ -181,287 +188,262 @@ func (b *dbBuffer) Write(
 	value float64,
 	unit xtime.Unit,
 	annotation []byte,
+	wopts WriteOptions,
 ) (bool, error) {
 	now := b.nowFn()
-	futureLimit := now.Add(1 * b.bufferFuture)
-	pastLimit := now.Add(-1 * b.bufferPast)
-	if !futureLimit.After(timestamp) {
-		return false, m3dberrors.ErrTooFuture
-	}
-	if !pastLimit.Before(timestamp) {
-		return false, m3dberrors.ErrTooPast
-	}
+	wType := b.resolveWriteType(timestamp, now, wopts)
+
+	if wType == ColdWrite {
+		if !b.coldWritesEnabled {
+			return false, m3dberrors.ErrColdWritesNotEnabled
+		}
+
+		if now.Add(-b.retentionPeriod).Truncate(b.blockSize).After(timestamp) {
+			return false, m3dberrors.ErrTooPast
+		}
 
-	bucketStart := timestamp.Truncate(b.blockSize)
-	idx := b.writableBucketIdx(timestamp)
-	if b.buckets[idx].needsReset(bucketStart) {
-		// Needs reset
-		b.DrainAndReset()
+		if !now.Add(b.futureRetentionPeriod).Truncate(b.blockSize).Add(b.blockSize).After(timestamp) {
+			return false, m3dberrors.ErrTooFuture
+		}
 	}
 
-	return b.buckets[idx].write(timestamp, value, unit, annotation)
+	blockStart := timestamp.Truncate(b.blockSize)
+	buckets := b.bucketVersionsAtCreate(blockStart)
+	b.putBucketVersionsInCache(buckets)
+	return buckets.write(timestamp, value, unit, annotation, wType)
```

Comment:
can we pass down writeOpts all the way down to bucket?

Reply:
nvm, please ignore that. I will pass down schema directly (so that we avoid allocating writeOpts for every write).